### PR TITLE
Add mechanism for user to protect label from recipe

### PR DESCRIPTION
**Pull request by jw3126**
@briochemc instead of being clever about discovering whether the unit is present in `xguide`, this adds an escape mechanism. What do you think?

### CODE REVIEW

**Comment by briochemc** [comment]
My opinion is that while it's nice to have the option, one should always show units on a plot of data that has units. So I don't think I would personally use that escape mechanism.

However, what I sometimes need is to show no label at all. This happens fairly often when there are subplots that share the same axis for example. So I think we need a mechanism for that. We could replicate the bhavior of the `label` keyword for the legend. I.e., if `attr[:xguide] == ""`, then we don't add the unit to it. So in short:
```julia
if haskey(attr, :xguide)
    if attr[:xguide] == ""
        # don't change xguide
    elseif # xguide does not already have a unit
        # append unit
    end
else
    # add the unit on its own
end

**Comment by jw3126** [comment]
So here is an example, why I don't want to search for the units. Say the user decides that `m s^-1` looks pretty ugly and wants `m/s` instead. Our search mechanism can't discover this and will add the ugly `m s^-1` anyway.

**Comment by jw3126** [comment]
The `xguide=""` idea is reasonable anyway.

**Comment by briochemc** [comment]
> So here is an example, why I don't want to search for the units. Say the user decides that `m s^-1` looks pretty ugly and wants `m/s` instead. Our search mechanism can't discover this and will add the ugly `m s^-1` anyway.

I agree this is ugly. But Plots and Unitful are not completely Unicode-ready anyway. E.g., `μs` will not display nicely with the GR backend for me. Note that Unitful should eventually work with Unicode superscripts ([hopefully soon for OSX and Linux](https://github.com/PainterQubits/Unitful.jl/pull/297)), and then it will look much better! (See `plot(rand(10)*u"μm/s")` example with `pyplot()` backend and my Unitful fork below.)

unicodeUnitfulPyPlot

For the mechanism you propose I think the name is quite obscure. I would suggest replacing it with a keyword maybe? What about `plot(y, forceylabel="m/s"`?

--- 

**EDIT**

FYI and for the record I realized unicode sort of works if you set
```julia
ENV["GKS_ENCODING"] = "utf-8"
```
before using Plots with the GR backend. However this does not work when saving to png, and I'm not sure it would work in the generated docs either. 

So what about setting the backend to pyplot? (That *might* also solve our cropping issue?)

**Comment by jw3126** [comment]
Maybe the name `Literally` is obscure? But the mechanic is julian IMO. Much more generic then adding more keyword arguments.

**Comment by briochemc** [comment]
Yes I'm not sure what would be a good name...

I'm also scared that this won't work if the recipe passes multiple times on the same axis. Could you test if it works with something like this input below?

```julia
x, y = [rand(10), rand(15), rand(20)]*u"m", [rand(10), rand(15), rand(20)]*u"g"
plot(x, y, xlabel=Literally("no touchy"))
```

**Comment by jw3126** [comment]
> 
> So what about setting the backend to pyplot? (That _might_ also solve our cropping issue?)

I am ready to do almost anything if it solves the cropping issue 😃

**Comment by briochemc** [comment]
> I am ready to do almost anything if it solves the cropping issue 😃

Haha go ahead if you want to try it :) But at the same time I think it is best to have the most minimalist examples (so just `using Plots` and no `pyplot()` involved) and since there is no rush and I asked on Slack maybe we can wait a little bit more?

**Comment by jw3126** [comment]
grafik

**Comment by jw3126** [comment]
Ok no `Literally`. I think the keyword argument variant would suffer the same problem?

**Comment by briochemc** [comment]
What about making your type a subtype of AbstractString (such that it prints) and not stripping it? (So that it stays protected regardless of the number of passes of the recipe)

**Comment by jw3126** [comment]
That is an interesting idea.

**Comment by briochemc** [comment]
OK I pulled this branch and will try to push a commit to it with the subtyping. 

On my local branch so far, I basically added what I think is the minimum `AbstractString` interface. I also added a string macro so that doing `xlabel=P"my label"` will assign a protected `"my label"` string and is maybe nicer than `"ProtecedString("my label)"` (at least it is shorter).

And as you can see yes, I am planning to change `Literally` to `ProtectedString` because I have an idea that you might like on the same principle for the next PR... To create a `UniftulString` type that is similarly protected, but that has not only the `content::String` but also the unit itself, so that we can perform checks and convert different units for the same axis on the fly, e.g., for something like
```julia
x, y = [rand(10)*u"m", rand(15)*100u"cm", rand(20)*u"m"], [rand(10), rand(15), rand(20)]*u"g"
plot(x, y, xunit=u"dm")
```
to work. In the (far) future this could also allow for different axes to be automatically created for figures of this type:
multiple_axes

File: src/UnitfulRecipes.jl [hunk 1]
```diff
@@ -86,24 +86,45 @@ function _resolve_axis!(attr, arr, T, axis)
...
+const S = ProtectedString
+Base.iterate(n::S) = iterate(n.content)
+Base.iterate(n::S, i::Integer) = iterate(n.content, i)
+Base.codeunit(n::S) = codeunit(n.content)
+Base.ncodeunits(n::S) = ncodeunits(n.content)
+Base.isvalid(n::S, i::Integer) = isvalid(n.content, i)
+Base.pointer(n::S) = pointer(n.content)
+Base.pointer(n::S, i::Integer) = pointer(n.content, i)
+# macro for easy-to-use interface?
+# i.e., so that `U"foo"` creates `ProtectedString("foo")`
```
Comment (briochemc) [inline src/UnitfulRecipes.jl]: I meant P"foo" here, corrected locally and will push after our next chat if that's OK 😃

**Comment by briochemc** [comment]
OK I think you should be happy with this one... It uses a `UnitfulString` to dispatch on whether or not to add the unit to the label, so no more checking the end of the current string of the label. 

It will also raise an error if the unit of the label and the unit of the data are not the same, although there is no such test. The thing is, if we do something like
```julia
plot(rand(10) * u"m")
plot!(rand(10) * u"s")
```
in the second line calling `!plot`, it seems that the `:yguide` attribute has disappeared, which I did not expect... 

Anyway, let me know what you think! BTW reminder: You can check the docs generated by this PR before merging [here](https://jw3126.github.io/UnitfulRecipes.jl/previews/PR16/)!

File: test/runtests.jl [hunk 2]
```diff
@@ -9,7 +9,15 @@ Attributes = Dict{Symbol, Any}
     ys = ys_val * m
     ys_ret = recipe!(attr, ys)
     @test ys_ret ≈ ys_val
-    @test attr[:yguide] == string(m)
+    @test attr[:yguide] == "m"
+
+    attr = Attributes(:yguide => P"hello")
+    recipe!(attr, ys)
+    @test attr[:yguide].content == "hello"
```
Comment (jw3126) [inline test/runtests.jl]:
```suggestion
    @test attr[:yguide] == UnitfulRecipes.ProtectedString("hello")
```
So it does not accidentally pick up the `ys` unit.

File: src/UnitfulRecipes.jl [hunk 3]
```diff
@@ -76,25 +77,70 @@ function _resolve_axis!(attr, arr, T, axis)
...
+    attr[key] = UnitfulString(string(u), u)
+    return attr
+end
+function append_unit_if_needed!(attr, key, label::UnitfulString, u)
+    (label.unit ≠ u) && error("The unit of $key has changed!")
+    return attr
+end
+function append_unit_if_needed!(attr, key, label::String, u)
+    if label ≠ ""
+        attr[key] = UnitfulString(string(label, " (", u, ")"), u)
```
Comment (jw3126) [inline src/UnitfulRecipes.jl:138]: In my experience `[kg]` is more common then `(kg)`, but I can live with both variants.

**Comment by jw3126** [comment]
Thanks a lot for all the good ideas and code contributions to this PR! I like it.

File: src/UnitfulRecipes.jl (hunk 3 above)
Comment (briochemc) [reply src/UnitfulRecipes.jl:138]: Oh yes sorry, I don't care either, it's just that I started by copying what I had before. Maybe we can have an enironment variable to set for either of the two?

Comment (jw3126) [reply src/UnitfulRecipes.jl:138]: I think we should stick to one instead of introducing this extra complexity. I am okay with `(kg)`.

**Comment by briochemc** [comment]
No worries, I'm adding a few examples and tests to improve coverage and I think I'll stop for a while after that 😃

File: src/UnitfulRecipes.jl (hunk 3 above)
Comment (briochemc) [reply src/UnitfulRecipes.jl:138]: OK no worries, I'll leave it as is and we can always change it later!

**Comment by briochemc** [comment]
OK I'm gonna push a commit with those tests and you will notice that a few of them are "escaped" (if that's the right term) because they currently fail, mostly for when data is supplied as vectors of vectors... I think these are edge cases that we can deal with later 😃

File: test/runtests.jl [hunk 4]
```diff
@@ -61,49 +63,56 @@ end
...
+            @test plot(x -> x^2, x*m)          isa Plots.Plot
         end
 
         @testset "Two arrays" begin
             @test plot(x*m, y*s)                    isa Plots.Plot
             @test plot(x*m, y*s, xlabel="x")        isa Plots.Plot
             @test plot(x*m, y*s, xlims=(-1,1))      isa Plots.Plot
             @test plot(x*m, y*s, xlims=(-1,1) .* m) isa Plots.Plot
-            @test plot(x*m, y*s, xunit=u"km")       isa Plots.Plot
+            dtype ≠ Symbol("Vectors of vectors") && @test plot(x*m, y*s, xunit=u"km") isa Plots.Plot
```
Comment (jw3126) [inline test/runtests.jl]:
```suggestion
            if dtype == Symbol("Vectors of vectors") 
                @test_broken plot(x*m, y*s, xunit=u"km") isa Plots.Plot
            else
                @test plot(x*m, y*s, xunit=u"km") isa Plots.Plot
            end
```
This clarifies the intention.

**Comment by jw3126** [comment]
Shall we merge it?
